### PR TITLE
Support server-side-rendered shadow roots

### PR DESCRIPTION
**Pull request by dvoytenko**
[Design doc](https://docs.google.com/document/d/1_tKo3g8XF1U65NxViW_lM4qA9JrwjXQZyeFx7wbE6rg/edit#)

### CODE REVIEW

File: src/preact/base-element.js [hunk 1]
```diff
@@ -238,20 +247,43 @@ export class PreactBaseElement extends AMP.BaseElement {
...
-        this.container_ = shadowRoot;
-
-        const shadowCss = Ctor['shadowCss'];
-        if (shadowCss) {
-          installShadowStyle(shadowRoot, this.element.tagName, shadowCss);
+        // Check if there's a pre-constructed shadow DOM.
+        let {shadowRoot} = this.element;
+        let container = shadowRoot && childElementByTag(shadowRoot, 'c');
+        if (container) {
+          toHydrate = true;
```
Comment (jridgewell) [inline src/preact/base-element.js:254]: I think this is going to cause every rerender to be a hydrate. Is that right?

Comment (dvoytenko) [reply src/preact/base-element.js:254]: This part of code executes only on first render. All other renders bypass this branch.

Comment (dvoytenko) [reply src/preact/base-element.js:254]: I found an easy way to test it too. Added tests. PTAL.

File: src/preact/base-element.js [hunk 2]
```diff
@@ -265,20 +273,43 @@ export class PreactBaseElement extends AMP.BaseElement {
...
         );
-        const shadowRoot = this.element.attachShadow({mode: 'open'});
-        this.container_ = shadowRoot;
-
-        const shadowCss = Ctor['shadowCss'];
-        if (shadowCss) {
-          installShadowStyle(shadowRoot, this.element.tagName, shadowCss);
+        // Check if there's a pre-constructed shadow DOM.
+        let {shadowRoot} = this.element;
+        let container = shadowRoot && childElementByTag(shadowRoot, 'c');
```
Comment (jridgewell) [inline src/preact/base-element.js:278]: Do we really need to create a new `<c>` element? Couldn't we just infer that an already created shadow root means that this was SSR'd?

Comment (dvoytenko) [reply src/preact/base-element.js:278]: I didn't find a stable enough way tbh. Here's what happens here. The shadow root looks like this

```
#shadow-root
  <style>...</style>
  <slot name="i-amphtml-svc"></slot>
  <c>
     <component-root>...</component-root>
  </c>
```

The `component-root` is whatever element that the hydration will go into. The `hydrate` API has to be called on the container. E.g.

```
hydrate(c, Component)
```

If I were to call `hydrate()` on a shadow root, it can get confused. E.g. if the root of a component was a `slot` element (e.g. in the passthrough mode), it'd think that `<slot name="i-amphtml-svc">` is the one to hydrate with.

So this is a very unstable situation. It's less likely to collide on `<style>`, of course. But any other children of shadow root would be up for grabs. Unfortunately, there's no API to say "hydrate only this node" - it only takes the whole container.

File: src/preact/index.js [hunk 3]
```diff
@@ -43,9 +43,18 @@ export function cloneElement(unusedElement, unusedProps, unusedChildren) {
 /**
  * @param {?PreactDef.VNode} vnode
  * @param {Node} container
+ * @param {?Node=} opt_replaceNode
  */
-export function render(vnode, container) {
-  preact.render(vnode, container, undefined);
+export function render(vnode, container, opt_replaceNode) {
```
Comment (samouri) [inline src/preact/index.js:48]: Is the addition of `opt_replaceNode` an unrelated change?

File: src/preact/base-element.js (hunk 2 above)
Comment (samouri) [reply src/preact/base-element.js:278]: Thank you for the thorough explanation! I'm surprised that Preact gets confused about which node to hydrate given multiple options

Comment (dvoytenko) [reply src/preact/base-element.js:278]: The first render/hydrate is most important when the framework is mounted. So I don't think it's that surprising give the current APIs. I think it'd have been relatively simple thing if APIs were different. I'll keep playing around to see if we can find a better hydration approach. But it's a tradeoff between trying to rely on some undocumented nuances of the implementation vs what API says.

File: src/preact/index.js (hunk 3 above)
Comment (dvoytenko) [reply src/preact/index.js:48]: Yeah. Noticed that we missed it at first.
